Source/SpatialGDK/Private/SpatialInterop.cpp

```diff
@@ -922,16 +929,146 @@ bool USpatialInterop::IsSingletonClass(UClass* Class)
+	{
+		return &(GSMD->stably_named_path_to_entity_id());
+	}
+
+	return nullptr;
+}
+
+void USpatialInterop::ReserveReplicatedStablyNamedActorChannel(USpatialActorChannel* Channel)
+{
+	ReplicatedStablyNamedActorQueue.Emplace(Channel->Actor, Channel);
+
+	if (CanSpawnReplicatedStablyNamedActors())
+	{
+		RegisterReplicatedStablyNamedActors();
+	}
+}
+
+void USpatialInterop::UnreserveReplicatedStablyNamedActor(AActor* Actor)
+{
+	ReplicatedStablyNamedActorQueue.Remove(Actor);
+}
+
+void USpatialInterop::AddReplicatedStablyNamedActorToGSM(const FEntityId& EntityId, AActor* Actor)
+{
+	PathNameToEntityIdMap* StablyNamedPathToEntityId = GetStablyNamedPathToEntityId();
+
+	if (StablyNamedPathToEntityId == nullptr)
+	{
+		return;
+	}
+
+	std::string Path = TCHAR_TO_UTF8(*Actor->GetPathName(Actor->GetWorld()));
+
+	// If the map already has the entity id, meaning the actor was already spawned, return early
+	if (StablyNamedPathToEntityId->count(Path))
+	{
+		return;
+	}
+
+	StablyNamedPathToEntityId->emplace(Path, EntityId.ToSpatialEntityId());
+
+	improbable::unreal::GlobalStateManager::Update Update;
+	Update.set_stably_named_path_to_entity_id(*StablyNamedPathToEntityId);
+
+	TSharedPtr<worker::Connection> Connection = SpatialOSInstance->GetConnection().Pin();
+	Connection->SendComponentUpdate<improbable::unreal::GlobalStateManager>(worker::EntityId((long)SpatialConstants::GLOBAL_STATE_MANAGER), Update);
+}
+
+void USpatialInterop::RegisterReplicatedStablyNamedActors()
+{
+	bCanSpawnReplicatedStablyNamedActors = true;
+	PathNameToEntityIdMap& StablyNamedPathToEntityId = *GetStablyNamedPathToEntityId();
+
+	for (const auto& Pair : ReplicatedStablyNamedActorQueue)
+	{
+		USpatialActorChannel* Channel = Pair.Value;
+		AActor* Actor = Pair.Key;
+		std::string Path = TCHAR_TO_UTF8(*Actor->GetPathName(Actor->GetWorld()));
+
+		// If the Actor was already spawned, just register the EntityId
+		if (StablyNamedPathToEntityId.count(Path))
+		{
+			Channel->RegisterEntityId(StablyNamedPathToEntityId[Path]);
+			continue;
+		}
+
+		// Otherwise, initiate the reservation (and thus creation) process
+		Channel->SendReserveEntityIdRequest();
+	}
+
+	ReplicatedStablyNamedActorQueue.Empty();
+}
+
+void USpatialInterop::DeleteIrrelevantReplicatedStablyNamedActors(const PathNameToEntityIdMap& StablyNamedPathToEntityId)
+{
+	for (const auto& Pair : StablyNamedPathToEntityId)
+	{
+		FString FullPath = UTF8_TO_TCHAR(Pair.first.c_str());
+		AActor* StableActor = FindObject<AActor>(GetWorld(), *FullPath);
+		worker::EntityId EntityId = Pair.second;
+
+		if (StableActor != nullptr && ReplicatedStablyNamedActorTimeoutMap.Find(EntityId) == nullptr)
+		{
+			TWeakObjectPtr<AActor> ActorPtr = StableActor;
+			FTimerHandle TimerHandle;
+			FTimerDelegate TimerCallback;
+			TimerCallback.BindLambda([ActorPtr, EntityId, this] {
+				ReplicatedStablyNamedActorTimeoutMap.Remove(EntityId);
+
+				if (AActor* Actor = ActorPtr.Get())
```

Comment:
Refactored this function slightly into two nested ifs, and one dereference of the WeakPtr.